ocamlPackages: update a bunch of libraries from the mirage universe

sternenseemann opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

* `Cstruct.len` has been finally deprecated in ctruct 6.0.1. A lot of libraries will have to be updated for this (see <https://github.com/ocaml/opam-repository/pull/19143>, <https://github.com/ocaml/opam-repository/pull/19142>).
* `tls` 0.14 breaks `paf`
* `paf` has a new release which supports `tls` 0.14 which breaks `git-paf` ~~which hasn't had a new release yet.~~
* `paf` >= 0.0.4 needs `letsencrypt` 0.3.0

More to come most likely, we are now waiting for a lot of `mirage-*` libraries to update as well as things like `conduit-lwt-unix`.

One thing to watch is also the use of `io-page-unix`. If nothing depends on it in nixpkgs anymore we can update `io-page`.

cc @vbgl for their information

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

sternenseemann committed:
ocamlPackages.letsencrypt: 0.2.5 -> 0.3.0

ocamlPackages.letsencrypt-dns: init at 0.3.0
ocamlPackages.letsencrypt-app: init at 0.2.5

https://github.com/mmaker/ocaml-letsencrypt/releases/tag/v0.3.0

sternenseemann committed:
ocamlPackages.cstruct: 6.0.0 -> 6.0.1

https://github.com/mirage/ocaml-cstruct/releases/tag/v6.0.1

sternenseemann committed:
ocamlPackages.tls: 0.13.2 -> 0.14.1

sternenseemann committed:
ocamlPackages.paf: 0.0.3 -> 0.0.5

ocamlPackages.paf-cohttp: init at 0.0.5
ocamlPackages.paf-le: init at 0.0.5

SuperSandro2000 commented on pkgs/development/ocaml-modules/cstruct/ppx.nix:
```diff
@@ -13,20 +13,8 @@ buildDunePackage {
-  # prevent ANSI escape sequences from messing up the test cases
-  # https://github.com/mirage/ocaml-cstruct/issues/283
-  patches = [
-    (fetchpatch {
```
Please remove the unused fetchpatch from inputs.

sternenseemann committed:
ocamlPackages.conduit*: 4.0.0 -> 4.0.1

sternenseemann committed:
ocamlPackages.git*: 3.4.0 -> 3.5.0

sternenseemann commented:
Still failing:

- [ ] jackline
- [ ] ocamlPackages.mirage-block-combinators
- [ ] ocamlPackages.mirage-block-ramdisk
- [ ] ocamlPackages.tezos-error-monad
- [ ] ocamlPackages.tezos-micheline
- [ ] ocamlPackages.tezos-protocol-environment-sigs
- [ ] ocamlPackages.tezos-rpc
- [ ] ocamlPackages.tezos-stdlib
- [ ] ocamlPackages.tls-async

tg-x commented:
#139644 also updates tls, i suggest closing that one and adding the x509 and macaddr update to this PR

sternenseemann commented:
Can try that later when I get home again, sounds good. Thanks for the heads up!

> On 28. Sep 2021, at 14:10, TG × ⊙ ***@***.***> wrote:
> 
> ﻿
> #139644 also updates tls, i suggest closing that one and adding the x509 and macaddr update to this PR
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

sternenseemann committed:
ocamlPackages.{ipaddr, macaddr}: 5.1.0 -> 5.2.0

https://github.com/mirage/ocaml-ipaddr/releases/tag/v5.2.0

Co-Authored-By: TG ⊗ Θ <*@tg-x.net>

sternenseemann committed:
jackline: unstable-2021-04-23 -> unstable-2021-08-10

sternenseemann committed:
ocamlPackages.mirage-block-combinators: patch for cstruct 6.0.0

The patch is a trivial search and replace of Cstruct.len ->
Cstruct.length otherwise the build will fail due to deprecation warnings.

sternenseemann commented:
`nixpkgs-review` is clean now.

tg-x commented:
OK, so looks like we could update x509 to 0.14.1 for now

sternenseemann committed:
ocamlPackages.x509: 0.14.0 -> 0.14.1

vbgl commented on pkgs/development/ocaml-modules/cstruct/ppx.nix:
```diff
@@ -13,20 +12,8 @@ buildDunePackage {
 
   minimumOCamlVersion = "4.07";
```
```suggestion
  minimalOCamlVersion = "4.08";
```

Build of `ocaml-ng.ocamlPackages_4_07.ppx_cstruct` fails with:

>   File "ppx/ppx_cstruct.ml", line 347, characters 27-36:
  Error: This expression has type 'a option
         but an expression was expected of type string

sternenseemann replied:
Done.

vbgl replied:
This was also a suggestion to fix the spelling…

sternenseemann replied:
Oh right, completely forgot that change happened. Maybe we should remove those treewide before 21.11 and make the old spelling a warning? Either way fixed.

vbgl commented:
Is there any value in the git history of this PR or can I squash it into a single commit (I did not test the intermediate states but I guess that there may be some broken packages in them)?

sternenseemann commented:
> Is there any value in the git history of this PR or can I squash it into a single commit (I did not test the intermediate states but I guess that there may be some broken packages in them)?

I believe a merge commit should make them atomic? But yes, there is most likely intermediate breakage.
